Brother MFC7860DW cups lpr wrapper drivers.

jraygauthier opened the pull request:
Also includes Brother generic lpr wrappers.

Use these on both my i686 and 64 bits machines.

Adapted from `*.deb` packages provided by brother.

codyopel commented:
From what I can tell you are using a printer specific driver, but it looks like Brother offers a 'Generic LPR Driver' that at a glance should support all current Brother models, and same for the cups wrapper.  It would be better to use these if they work, rather than including a separate package from each printer model unless necessary.  I will look into it further if I get a chance.

jraygauthier commented:
Yes indeed, the generic lpr / cup wrapper are:

~~~
brgenml1lpr = callPackage ../misc/cups/drivers/brgenml1lpr {};
brgenml1cupswrapper = callPackage ../misc/cups/drivers/brgenml1cupswrapper {};
~~~

while the specific ones are:

~~~
cupswrapperMFC7860DW = callPackage ../misc/cups/drivers/cupswrapperMFC7860DW {};
brmfc7860dwcups = callPackage ../misc/cups/drivers/brmfc7860dwcups {};
mfc7860dwlpr = callPackage ../misc/cups/drivers/mfc7860dwlpr {};
~~~

where `cupswrapperMFC7860DW` seems to be literally a prebuilt version of `brmfc7860dwcups`.

The generic version seem to work pretty much as well as the specific ones. Maybe the added
value of the specific one is their cups filters. 

It was unclear to me what was the policy
about printer drivers on nix. So I though maybe these could be pretty good examples
for other people adding their own brother specific drivers.

jraygauthier commented:
Here's what I found in term of embedded string binary patching:

<http://ohnopub.net/~ohnobinki/bin_replace_string/>

Limitation is that the string be shorter.

The other way around was to duplicate the `/etc/*` configuration file generator
so that doing the following (as documented in comments) is possible in one's
nixos expression to automatically generate new configuration file under `/usr/*`.

~~~
environment.usr."local/Brother/Printer/MFC7860DW/inf/brMFC7860DWfunc".source =
"${pkgs.mfc7860dwlpr}/usr/local/Brother/Printer/MFC7860DW/inf/brMFC7860DWfunc";
~~~

This thing has not been yet submitted for review. I believe it wouldn't be accepted
anyway as it kind of defeat the (controlled) immutability principle of nixos.

codyopel commented:
So do the generic drivers work just fine or are the model specific ones still needed, it sounds like from what your stating that the generics work.  It would be cleaner if we don't have to specify a specific driver and instead could enable brother printers as a whole.

For the config I'll have to look into what is going on, but typically it is best to avoid having to put anything in `/usr` if at all possible.  It defeats the purpose of the `/nix/store`.  I am trying to solve a similar problem with the nvidia drivers at the moment.

jraygauthier commented:
Hi again @codyopel ,

yes generic driver work fine. These are the ones I use right now. 

I read the following blog post about patching issues with hardcoded paths
in executable: <http://blog.lastlog.de/posts/creative_packaging_in_nixpkgs/>.

I think this is the way to go.

joachifm commented:
Ping?

jraygauthier committed:
brgenml1cupswrapper: init at 3.1.0-1

Brother generic cups wrapper over lpr driver
`brgenml1lpr` init at same version.

jraygauthier commented:
Hi @joachifm. Removed specific MFC drivers and kept only the generic ones as suggested. Also
rebased on top of latest `master` branch.

joachifm commented on pkgs/misc/cups/drivers/brgenml1cupswrapper/default.nix:
```diff
@@ -0,0 +1,108 @@
+
+  name = "brgenml1cupswrapper-3.1.0-1";
+  src = fetchurl {
+    url = "http://download.brother.com/welcome/dlf101125/brgenml1cupswrapper-3.1.0-1.i386.deb";
```
Re-using the name attribute here makes bumps easier, only need to change the version in one place.

joachifm commented on pkgs/misc/cups/drivers/brgenml1cupswrapper/default.nix:
```diff
@@ -0,0 +1,108 @@
+    WRAPPER=opt/brother/Printers/BrGenML1/cupswrapper/brother_lpdwrapper_BrGenML1
+    PAPER_CFG=opt/brother/Printers/BrGenML1/cupswrapper/paperconfigml1  
+
+    #chgrp -R root opt
```
Remove this?

joachifm commented on pkgs/misc/cups/drivers/brgenml1lpr/default.nix:
```diff
@@ -0,0 +1,92 @@
+    platforms = stdenv.lib.platforms.linux;
+    license = stdenv.lib.licenses.gpl2Plus;
+  };
+}
```
Newline error

joachifm commented:
I encourage you to add yourself to the list of maintainers. Otherwise, this looks good to me, but because I don't know anything about printer drivers, I'd like to wait just a little bit longer for feedback, e.g., from @codyopel before merging, hope that's ok.

jraygauthier commented:
@joachifm: Updated with suggested changes.

joachifm commented on pkgs/misc/cups/drivers/brgenml1cupswrapper/default.nix:
```diff
@@ -0,0 +1,107 @@
+  '';
+
+  buildInputs = [ cups perl brgenml1lpr ];
+  buildPhase = ''
```
This can be shortened to `buildPhase = ":"` or you could set `phases` directly. Not that important.

jraygauthier replied:
done (the `":"` solution).

joachifm commented on pkgs/misc/cups/drivers/brgenml1cupswrapper/default.nix:
```diff
@@ -0,0 +1,107 @@
+    tar xfvz data.tar.gz
+  '';
+
+  buildInputs = [ cups perl brgenml1lpr ];
```
Where is perl being used?

jraygauthier replied:
`perl` is needed at runtime as interpreter of `paperconfigml1` and
`brother_lpdwrapper_BrGenML1`(introduced at automatic interpreter patching step).

joachifm commented on pkgs/misc/cups/drivers/brgenml1lpr/default.nix:
```diff
@@ -0,0 +1,93 @@
+     -  The `setupPrintcap` has totally no use in our context.
+*/
+
+let myPatchElf = (file: 
```
This can be simplified; as far as I can tell, the only difference between these branches are the suffix on ld-linux.so; it is cleaner to do something like `patchelf --set-interpreter ${stdenv.glibc}/lib/ld-linux${optionalString stdenv.is64bit "x86-64"}.so.2 ${file}`

jraygauthier replied:
Done. Good idea. Much tidier this way.

joachifm commented on pkgs/misc/cups/drivers/brgenml1lpr/default.nix:
```diff
@@ -0,0 +1,93 @@
+    tar xfvz data.tar.gz
+  '';
+
+  buildInputs = [ cups perl glibc ghostscript which makeWrapper];
```
`makeWrapper` could be moved to `nativeBuildInputs`, possibly perl as well?

jraygauthier replied:
done for `makeWrapper`. `perl` is however needed at runtime as interpreter of `filter_BrGenML1` (introduced at automatic interpreter patching step).

joachifm commented:
I just had to take a second look at this, so a couple more nitpicks ... I'll merge this once they have been addressed. This PR has been open long enough, I think ...

jraygauthier commented:
@joachifm: Updated patch with suggested changes.

joachifm commented:
:beers: :

joachifm commented:
Ach, sorry, ONE last thing: please rename the commit subject to `pkgname: init at version`.

jraygauthier commented:
done.

joachifm commented:
Thanks
